### PR TITLE
Introduce RetrySettings support for Hashicorp Vault

### PR DESCRIPTION
**Pull request by 12345ieee**
## Problem Statement

Hashicorp vault sdk supports retry attempts/interval, but those are not bound to the `{Cluster,}SecretStore` ones, bind them.

More about the actual use case, I have many secrets set at a 1m interval, and I see rare failures in syncing due to vault internal errors.
Those failures are transmitted to Argocd via the ES status and then to humans via email, I want to see if setting the retry amount to 1 or 2 (up from effectively 0) solves it.

## Related Issue

I opened no issue, tell me if I need to.

## Proposed Changes

Make it so the `SecretStoreSpec.retrySettings` are actually passed to the vault provider implementation.
Also update the documentation.

It's my first time writing anything in go, it was not easy to set everything up and get this done, please advise if the code has issues.
I've tested the image in a real cluster for a bit, everything worked, adding a new e2e test is outside my current ability.

## Checklist

- [x] I have read the [contribution guidelines](https://external-secrets.io/latest/contributing/process/#submitting-a-pull-request)
- [x] All commits are signed with `git commit --signoff`
- [x] My changes have reasonable test coverage
- [x] All tests pass with `make test`
- [x] I ensured my PR is ready for review with `make reviewable`

### CODE REVIEW

File: pkg/provider/vault/vault.go
Comment (shuheiktgw) [inline pkg/provider/vault/vault.go:249]: I'm not too sure if we can or should pass `retrySettings` here, so I need advise on this point from others 🙇

**Review by shuheiktgw — COMMENTED** [review]
I left some comments, so please take a look at them! In addition to them, I believe we need to a unit test to very the retrySettings to merge this PR🙏

File: pkg/provider/vault/vault.go
Comment (12345ieee) [reply pkg/provider/vault/vault.go:288]: This was explicitly set to 0 (instead of the default of 2) in abec2a64cc236451c , it seems impolite to reverse the decision in an unrelated PR.

File: pkg/provider/vault/vault.go
Comment (12345ieee) [reply pkg/provider/vault/vault.go:293]: I'm not sure what the policy of this project is regarding breaking userspace code.

Technically a user could have had garbage in that field and it'd have worked till now, I'd expect the error reporting to start form the next minor or next major release, and be tracked separately.

Of course this is a 0.x.y, so maybe you guys want stuff to break if it's broken, just tell me your preferences and I'll fix it.

File: docs/contributing/devguide.md
Comment (12345ieee) [reply docs/contributing/devguide.md:50]: I've tried running `make docker.build` by itself and it didn't work, I've looked at the code and found out that when it's called from other build scripts those values are passed, I passed them and it worked.

I've followed the blames, the parameters passed changed across the years, thought it was just stale docs, if you can run `make docker.build` from a clean repo and it works let me know, I'll retry.

File: pkg/provider/vault/vault.go
Comment (12345ieee) [reply pkg/provider/vault/vault.go:249]: The generator spec at https://external-secrets.io/v0.8.5/api/generator/vault/ doesn't have it, nor do the other generators.

If it's desired it should IMHO be a separate PR, that I may be able to complete myself.

**Comment by 12345ieee** [comment]
> I left some comments, so please take a look at them! In addition to them, I believe we need to a unit test to very the retrySettings to merge this PR🙏

Thank you very much @shuheiktgw , I've left answers.
I didn't know go and I've learned just enough of it to contribute here and it was already a challenge, I agree on the need for a unit test, if everything else looks ok I'll try to produce one, it's a bit over my skill set at the moment.

File: docs/contributing/devguide.md
Comment (shuheiktgw) [reply docs/contributing/devguide.md:50]: `make docker.build` works fine in my local environment (M1 Mac, Docker version 24.0.2, Go version 1.19.3). Would you share the error message you received when you ran the command?

File: pkg/provider/vault/vault.go
Comment (shuheiktgw) [reply pkg/provider/vault/vault.go:288]: Hmm, if it is intentional, I'm fine with it 🙂

File: pkg/provider/vault/vault.go
Comment (shuheiktgw) [reply pkg/provider/vault/vault.go:293]: Ah, I see what you are trying to achieve here! Thank you for your explanation. The current implementation makes it hard for users to notice their misconfiguration, so I prefer simply returning the error (+ I believe we should have a validating using webhook, but that is not within the scope of this PR). But I understand it is a breaking change, so I'd like to know what other maintainers think 🙏  @moolen @gusfcarvalho

File: docs/contributing/devguide.md
Comment (12345ieee) [reply docs/contributing/devguide.md:50]:
```
$ LANG=C make docker.build 
09:13:24 [ OK ] Finished generating deepcopy and crds
09:13:24 [ .. ] go build amd64
CGO_ENABLED=0 GOOS=linux GOARCH=amd64 \
	go build -o 'bin/external-secrets-linux-amd64' main.go
09:13:24 [ OK ] go build amd64
09:13:24 [ .. ] go build arm64
CGO_ENABLED=0 GOOS=linux GOARCH=arm64 \
	go build -o 'bin/external-secrets-linux-arm64' main.go
09:13:25 [ OK ] go build arm64
09:13:25 [ .. ] docker build
echo docker build -f Dockerfile .  -t ghcr.io/external-secrets/external-secrets:v0.9.1-17.g8c8c17e0
docker build -f Dockerfile . -t ghcr.io/external-secrets/external-secrets:v0.9.1-17.g8c8c17e0
docker build -f Dockerfile .  -t ghcr.io/external-secrets/external-secrets:v0.9.1-17.g8c8c17e0
Sending build context to Docker daemon  292.9MB
Step 1/6 : FROM gcr.io/distroless/static
 ---> 961207f62413
Step 2/6 : ARG TARGETOS
 ---> Using cache
 ---> dec1fd805da3
Step 3/6 : ARG TARGETARCH
 ---> Using cache
 ---> 3c50e8bb496e
Step 4/6 : COPY bin/external-secrets-${TARGETOS}-${TARGETARCH} /bin/external-secrets
COPY failed: file not found in build context or excluded by .dockerignore: stat bin/external-secrets--: file does not exist
make: *** [Makefile:243: docker.build] Error 1
```
As you see, it breaks because `${TARGETOS}` and `${TARGETARCH}` are empty, so I gave them.
You may have them set in your shell config.

Comment (shuheiktgw) [reply docs/contributing/devguide.md:50]: I briefly looked into it and noticed `TARGETOS ` and `TARGETARCH` are arguments automatically set by BuildKit. Would you check if you enabled it in your environment? 

- https://docs.docker.com/engine/reference/builder/#automatic-platform-args-in-the-global-scope
- https://docs.docker.com/build/buildkit/

Comment (12345ieee) [reply docs/contributing/devguide.md]: Mystery revealed, I'm on version 20.10, if I use `DOCKER_BUILDKIT=1 make docker.build` it works.
Had no idea they were changing the build backed.

Fixed, but I could also drop the cange altogether if desired.

File: docs/contributing/devguide.md
Comment (shuheiktgw) [inline docs/contributing/devguide.md:50]: It would be better if we set `DOCKER_BUILDKIT=1` in the Makefile since engineers don't have to care about it. Also, you are right and the `IMG` actually does nothing, and we can just remove it 🙂 

Those changes are not directly related to the PR itself, so you can just revert the changes on this file if you don't want to work on the comment above!

File: docs/contributing/devguide.md
Comment (12345ieee) [reply docs/contributing/devguide.md]: Fixed all around.

File: pkg/provider/vault/vault.go
Comment (shuheiktgw) [reply pkg/provider/vault/vault.go]: I'm asking for advise from other contributors on this point!
https://kubernetes.slack.com/archives/C047LA9MUPJ/p1691538211317759

Comment (shuheiktgw) [reply pkg/provider/vault/vault.go]: No worries 🙂 It seems we agree that we can just return the error, so let's go with it!

**Comment by 12345ieee** [comment]
All the comments are resolved, now that I've figured out how go tests work I've added a retryInterval validation test.

File: pkg/provider/vault/vault_test.go
Comment (shuheiktgw) [inline pkg/provider/vault/vault_test.go:253]: Would you add one more test case for valid retry settings?

**Review by shuheiktgw — CHANGES_REQUESTED** [review]
I left one comment but overall it looks good to me 🙂

File: pkg/provider/vault/vault_test.go
Comment (12345ieee) [reply pkg/provider/vault/vault_test.go:253]: I need help for that, the code would be:
```go
		"ValidRetrySettings": {
			reason: "Should return a Vault provider with custom retry settings",
			args: args{
				store: makeSecretStore(func(s *esv1beta1.SecretStore) {
					s.Spec.RetrySettings = &esv1beta1.SecretStoreRetrySettings{
						MaxRetries:    pointer.To(int32(3)),
						RetryInterval: pointer.To("10m"),
					}
				}),
				ns:   "default",
				kube: clientfake.NewClientBuilder().Build(), // TODO here I need some magic instead of an empty object
			},
			want: want{
				err: nil,
			},
		},
```
but I need to know how to create a default but working kclient, or I get null pointers.
I don't want to copy it from another test or I'll introduce unrelated cross failures.

Comment (shuheiktgw) [reply pkg/provider/vault/vault_test.go:253]: Setting `corev1` and `newClientFunc` should do the work 🙂 

```
			args: args{
				store: makeSecretStore(func(s *esv1beta1.SecretStore) {
					s.Spec.RetrySettings = &esv1beta1.SecretStoreRetrySettings{
						MaxRetries:    pointer.To(int32(3)),
						RetryInterval: pointer.To("10m"),
					}
				}),
				ns:            "default",
				kube:          clientfake.NewClientBuilder().Build(),
				corev1:        utilfake.NewCreateTokenMock().WithToken("ok"),
				newClientFunc: fake.ClientWithLoginMock,
			},
```

Comment (12345ieee) [reply pkg/provider/vault/vault_test.go:253]: Here we go.

**Comment by shuheiktgw** [comment]
@12345ieee Thank you! It looks good to me 🙂 Would you resolve the conflict? Then I'll wait until I receive one more approval from other maintainers.

**Comment by 12345ieee** [comment]
Done.

**Comment by 12345ieee** [comment]
@shuheiktgw is there any chance this could make it into 0.9.4?
So that I can decide if I wait a few more days for an official release or just roll it as a custom image in my clusters ASAP.

**Comment by shuheiktgw** [comment]
Sorry, I've asked for a review again!
https://kubernetes.slack.com/archives/C047LA9MUPJ/p1693009454748039

**Review by moolen — APPROVED** [review]
code-wise LGTM, thanks for adding the retrySettings!
I'll need to wait for e2e tests to finish :hourglass_flowing_sand:

**Pull request merged**
